Source/Fuse.Scripting.JavaScript/Tests/Model/Model.Test.uno

```diff
@@ -0,0 +1,581 @@
+			using (var root = TestRootPanel.CreateWithChild(e))
+			{
+				root.StepFrameJS();
+				Assert.AreEqual("kaka", e.p.Value);
+
+				root.StepFrameJS();
+				Assert.AreEqual("yay!", e.t.Value);
+
+				e.changePromise.Perform();
+				root.StepFrameJS();
+				Assert.AreEqual("yay!", e.t.Value);
+
+				e.resolveNow.Perform();
+				root.StepFrameJS();
+				Assert.AreEqual("hoho!", e.t.Value);
+			}
+		}
+
+		[Test]
+		public void EmptyList() {
+			var e = new UX.Model.EmptyList();
+			using (var root = TestRootPanel.CreateWithChild(e)) {
+				var oc = e.collector1.Children;
+				var poc = e.collector2.Children;
+				root.StepFrameJS();
+				Assert.AreEqual("1,2,3", e.oc.JoinValues());
+				Assert.AreEqual("3,4,5", e.poc.JoinValues());
+				Assert.AreEqual(4, oc.Count);
+				Assert.AreEqual(4, poc.Count);
+				e.callEmpty.Perform();
+				e.callEmptyPromise.Perform();
+				root.StepFrameJS();
+				Assert.AreEqual("", e.oc.JoinValues());
+				Assert.AreEqual("", e.poc.JoinValues());
+				Assert.AreEqual(1, oc.Count);
+				Assert.AreEqual(1, poc.Count);
+			}
+		}
+
+		[Test]
+		public void MultiCounter() {
+			var e = new UX.Model.MultiCounter();
+			using (var root = TestRootPanel.CreateWithChild(e)) {
+				root.StepFrameJS();
+				var oc = e.counterCollector.Children;
+				Assert.AreEqual(2, oc.Count);
+
+				var counter1 = oc[1] as UX.Model.Counter;
+				Assert.AreNotEqual(null, counter1);
+
+				Assert.AreEqual("0", counter1.counter.Value);
+
+				counter1.callIncrement.Perform();
+				root.StepFrameJS();
+				Assert.AreEqual("1", counter1.counter.Value);
+
+				e.callAddCounter.Perform();
+				root.StepFrameJS();
+				Assert.AreEqual(3, oc.Count);
+
+				var counter2 = oc[2] as UX.Model.Counter;
+				Assert.AreNotEqual(null, counter2);
+
+				Assert.AreEqual("0", counter2.counter.Value);
+
+				counter2.callIncrement.Perform();
+				root.StepFrameJS();
+				Assert.AreEqual("1", counter2.counter.Value);
+
+				counter2.callIncrement.Perform();
+				root.StepFrameJS();
+				Assert.AreEqual("2", counter2.counter.Value);
+
+			}
+		}
+
+		[Test]
+		public void MutatePages()
+		{
+			var e = new UX.Model.MutatePages();
+			using(var root = TestRootPanel.CreateWithChild(e))
+			{
+				root.StepFrameJS();
+				var children = ChildrenOfType<Visual>(e.navigator);
+				Assert.AreEqual(1, children.Count);
+				Assert.OfType<UX.Model.MainPage>(children[0]);
+
+				e.pushPage.Perform();
+				root.StepFrameJS();
+
+				var childrenAfterPush = ChildrenOfType<Visual>(e.navigator);
+				Assert.AreEqual(2, childrenAfterPush.Count);
+				Assert.OfType<UX.Model.MainPage>(childrenAfterPush[0]);
+				Assert.OfType<UX.Model.DetailPage>(childrenAfterPush[1]);
+
+				e.popPage.Perform();
+				root.StepFrameJS();
+
+				Assert.OfType<UX.Model.MainPage>(e.navigator.Active);
+			}
+		}
+
+		List<T> ChildrenOfType<T>(Visual n) where T : Node
+		{
+			var l = new List<T>();
+			for (int i=0; i < n.Children.Count; ++i)
+			{
+				var m = n.Children[i] as T;
```

Comment:
Please do:
```uno
foreach (var child in n.Children)
{
	var m = child as T;
	[...]
```

This way, we don't have to bend over backwards to support indexing, which isn't needed in this case.